### ISSUE TITLE
Default PowerShell Terminal settings on Windows break certain paths

### ISSUE THREAD

**Issue by svArtist** [microsoft/vscode#138526]
Issue Type: <b>Bug</b>

The default "Open in  integrated Terminal" command, as used in the workspace -> right-click on folder context menu, still doesn't work with paths that PowerShell might interpret as patterns (such as `D:\[Projects]\[JS]\Web Crawler\HeadlessChrome`) instead of literal paths.
In this situation, we _know_ that we have a literal path, so the command should use the according switch:
`Set-Location -literalpath "${...}"` or `CD -literalpath "${...}"`
instead of just
`Set-Location "${...}"` or `CD "${...}"`
I don't know how it's actually implemented; I tried navigating this repository, but couldn't find the relevant resource; it's just so overwhelming when you don't know how something as big as this is organized 😅
It seems, anyway, like there should be an easy fix to be made somewhere.
Tested on Stable and Insiders build

PS: 
Possibly https://github.com/microsoft/vscode/blob/af376e98a8214d52cbce69fac6765dca8bea1b09/src/vs/workbench/contrib/debug/node/terminals.ts#L121 ?

![image](https://user-images.githubusercontent.com/13470026/144885531-6d5f9435-5599-4dfb-9a90-e3bb5d9f8226.png)

VS Code version: Code - Insiders 1.63.0-insider (d09289a2b6dd0e56b2a264a8e9380685498e10d6, 2021-12-04T02:05:02.410Z)
OS version: Windows_NT x64 10.0.19043
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|AMD FX(tm)-6300 Six-Core Processor              (6 x 3490)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: enabled_on<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|31.95GB (19.33GB free)|
|Process Argv|--crash-reporter-id 3a44598a-df18-499c-ab44-e0e055e2f4e2 --crash-reporter-id 3a44598a-df18-499c-ab44-e0e055e2f4e2|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (14)</summary>

Extension|Author (truncated)|Version
---|---|---
vsc-export|asl|1.2.0
code-gnu-global|aus|0.2.2
vscode-icontheme-nomo-dark|be5|1.3.6
working-sets|ber|1.4.0
jshint|dba|0.11.0
open-files|Div|0.0.4
file-icons|fil|1.0.29
beautify|Hoo|1.5.0
brackets-keybindings|ms-|0.1.1
cpptools|ms-|1.7.1
vscode-opened-editors|Nic|0.1.3
tabster|rib|0.0.6
reopenclosedtab|uyi|1.1.0
vscode-backup|Wes|0.0.6

</details>

**Comment by Tyriar** [comment]
Can you show a http://gifcap.dev/ of the problem?

**Comment by svArtist** [comment]
> Can you show a http://gifcap.dev/ of the problem?

Not too happy with that site: Seems like it doesn't support custom regions or tooltips/layered windows.
But if it's a GIF you want, a GIF you shall have:
![Capture 2021-12-06 22-00-19 optimized](https://user-images.githubusercontent.com/13470026/144922633-79d0c2ca-854d-4b25-91b9-d1abed6d5359.gif)

**Comment by Tyriar** [comment]
@joaomoreno FYI gifcap feedback

**Comment by Tyriar** [comment]
I can reproduce, this seems specific to Windows PowerShell. Since it works in cmd and powershell 6+, I think this is an issue with Windows PowerShell and the fix is to update PowerShell.

**Comment by svArtist** [comment]
The _Windows PowerShell_ as part of Windows is 5.1 (officially the final version), which is the default for most Windows users who haven't installed the independent _PowerShell_.
It seems harsh to discount the default case, especially since the fix would be so simple as to use the appropriate switch intended for these situations.
If VS Code will use _Windows Powershell_ as a default, it should employ the correct arguments, IMO

**Comment by toshafaye** [comment]
Tosha Wade

**Comment by Tyriar** [comment]
@svArtist we do use the correct arguments afaik, I believe it's a powershell bug. You should complain to the powershell team.

**Comment by svArtist** [comment]
I'm sorry I keep nagging, but it's not a bug. Square brackets in the `path` argument are interpreted as a "wildcard" pattern. That is the desired behavior by default, and still is for 7.2: https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.management/set-location?view=powershell-7.2#parameters
If you want to pass in a path and not have it interpreted as a wildcard pattern, that's literally what the `-literalpath` switch is there for.

![Capture 2021-12-07 21-48-19](https://user-images.githubusercontent.com/13470026/145104535-160d9eb3-d203-4c94-9357-1da3feb7763c.png)

Actually... Can you point me to the relevant code that sets up the arguments for opening a path in PS in this repo? I wonder how it manages to work in PS 6+.
I can't see the arguments in the command line, the magic must be happening after the terminal was launched.

Anyway; clearly, the implementation for _Windows PowerShell_ isn't working, and WPS is no longer in active development; the version that ships with Windows is _final_.
There is a TODO here, and it's either fixing how the CWD is set, or dropping support for WPS.
(Unless, of course, the means by which the CWD is set is controlled by a 3rd party solution)

**Comment by Tyriar** [comment]
The pty gets spawned here:

https://github.com/microsoft/vscode/blob/950e89beacff64928098d513b590a80401f42e2e/src/vs/platform/terminal/node/terminalProcess.ts#L249

The cwd of the shell is setup by setting the current directory on the process:

https://github.com/microsoft/node-pty/blob/256c59cd1adfb71053d8a79ad46c833b3e1915d8/src/win/conpty.cc#L361

Old PowerShell is handling it incorrectly since all other shells work fine. I know it's final, it sucks that this is a problem, but I don't think it's possible for us to fix it.

**Comment by svArtist** [comment]
Thank you for being so helpful and patient with me ^^
It is inconvenient that WPS is the outlier here, failing to handle this case.
It's certainly possible over all, I'd like to suggest that the question is how much of an exception you're comfortable with making.
I see that there's already a distinction being made between WPS and PS in some places; if only to provide a different icon.
It might be worth splitting up the two completely, and in the case of WPS, using something like `powershell -noexit -command "Set-location -LiteralPath '${cwd}'"` to spawn the console when there's a custom CWD (or maybe only after checking if that path does contain problem cases like square brackets)

I use the above (well, with `${workspaceFolder}`) to spawn in my problematic workspace folders, but it looks like there's no way to use the path of a file or folder selected in the workbench explorer in a custom terminal profile setting, and one might maybe argue that there shouldn't.

All things considered, can I not persuade you to reconsider the issue? :)

**Comment by Tyriar** [comment]
> using something like powershell -noexit -command "Set-location -LiteralPath '${cwd}'" to spawn the console

It's not that simple unfortunately, what if the powershell is being passed arguments? What if those arguments are escaped? There were some ongoing issues we had around escaping arguments and I'd rather not reopen that. Additionally, we try to avoid shell-specific logic as it's difficult to test and adds cruft to the code, also [Windows PowerShell is essentially deprecated](https://devblogs.microsoft.com/powershell/powershell-6-0-roadmap-coreclr-backwards-compatibility-and-more/#future-of-windows-powershell) (this is definitely a lower priority bug).

You'd be better off just not using the `[folder]` pattern.

**Comment by svArtist** [comment]
I see. I was only thinking of spawning the console "in isolation", then applying any other requirements, but I guess that's not always enough, even if such an exception were to be hard-coded in.
Maybe there should be an unmistakable deprecation warning somewhere. (The current implementation of WPS already mentions PS Core, but that phrasing doesn't convey the true nature of deprecation)

Thank yo so much for taking your time and explaining the state of things, I'm sure it will be helpful to others, too!
